back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb

```diff
@@ -4,256 +4,185 @@ module UserCustomFields
+              CustomField.find(params[:custom_field_id])
+            elsif (key = custom_field_key_from_path)
+              CustomField.find_by(key: key)
+            else
+              raise ActiveRecord::RecordNotFound
+            end
+        end
 
-          serie = users
-            .where(registration_completed_at: @start_at..@end_at)
-            .group("custom_field_values->'education'")
-            .order(Arel.sql("custom_field_values->'education'"))
-            .count
-          serie['_blank'] = serie.delete(nil) || 0 unless serie.empty?
+        def custom_field_key_from_path
+          request.path.split('/').last
+                 .match(/^users_by_(?<key>gender|education|birthyear|domicile)/)&.[](:key)
+        end
 
-          serie
+        def user_counts
+          @user_counts ||= count_users_by_custom_field(find_users, custom_field)
         end
 
-        def users_by_education
-          render json: { series: { users: users_by_education_serie } }
+        def count_users_by_custom_field(users, custom_field)
+          field_values = select_field_values(users, custom_field)
+
+          # Warning: The method +count+ cannot be used here because it introduces a SQL syntax
+          # error while rewriting the SELECT clause. This is because the 'field_value' column is a
+          # computed column and it does not seem to be supported properly by the +count+ method.
+          counts = field_values.order('field_value')
+                               .group('field_value')
+                               .select('COUNT(*) as count')
+                               .to_a.pluck(:field_value, :count).to_h
+
+          counts['_blank'] = counts.delete(nil) || 0
+          counts
         end
 
-        def users_by_education_as_xlsx
-          xlsx = XlsxService.new.generate_field_stats_xlsx users_by_education_serie, 'education', 'users'
-          send_data xlsx, type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', filename: 'users_by_education.xlsx'
+        # Returns an ActiveRecord::Relation of all the custom field values for the given users.
+        # It returns a view (result set) with a single column named 'field_value'. Essentially,
+        # something that looks like:
+        #   SELECT ... AS field_value FROM ...
+        #
+        # Each user results in one or multiple rows, depending on the type of custom field.
+        # Custom fields with multiple values (e.g. multiselect) are returned as multiple rows.
+        # If the custom field has no value for a given user, the resulting row contains NULL.
+        def select_field_values(users, custom_field)
+          case custom_field.input_type
+          when 'select', 'checkbox', 'number'
+            users.select("custom_field_values->'#{custom_field.key}' as field_value")
+          when 'multiselect'
+            users.joins(<<~SQL.squish).select('cfv.field_value as field_value')
+              LEFT JOIN (
```

Comment:
Slightly concerned that this query will be slow. Don't know the numbers involved for our biggest datasets.